File: javascript/src/websocket.js
const WebSocket = require('ws')

const config = require('./config.js')

/**
 * A WebSocket handler that parses messages from the server and spits them out accordingly
 */
class WebSocketHandler {
  /**
   * Construct a websocket handler
   * @param {Player} player The player object
   */
  constructor(player) {
    this.player = player
    this.ws = new WebSocket(`ws://${config.SERVER_HOST}/ws`, { headers: player.headers })

    this.ws.on('open', this.handleOpen.bind(this))
    this.ws.on('message', this.handleMessage.bind(this))
    this.ws.on('error', this.handleError.bind(this))
    this.ws.on('close', this.handleClose.bind(this))
  }

  /**
   * WebSocket handler for open events
   */
  handleOpen() {
    console.log(`Connected to ${config.SERVER_HOST}`)
  }

  /**
   * WebSocket event handler for message events
   * @param {String} msg The message from the server
   */
  handleMessage(msg) {
    const payload = JSON.parse(msg)
    const eventType = payload['type']
    const data = payload['data']

    // TODO: Push type and data into an event parser
    // TOOD: Make constants for each event type
    if (eventType === "game_message") {
      console.log(`[SERVER] ${data['msg']}`)
    } else {
      console.log(`[SERVER] (${eventType}) ${JSON.stringify(data)}`)
    }
  }

  /**
   * WebSocket event handler for error events
   * @param {Number} code The error code if available
   * @param {String} reason The reason if given
   */
  handleError(code, reason) {
    console.error(`[ERROR] ${code} ${reason}`)
  }

  /**
   * WebSocket event handler for close events
   */
  handleClose() {
    console.warn(`[WS] WebSocket connection closed`)
  }
}

module.exports = {
  WebSocketHandler: WebSocketHandler
}

File: python/src/websocket.py
import json

from aiohttp import ClientSession
from aiohttp import WSMsgType
from colorama import Fore

from config import SERVER_HOST


WS_URI: str = f'https://{SERVER_HOST}/ws'

GAME_MESSAGES = (
    'game_started',
    'game_message',
    'game_paused',
    'game_unpaused',
    'game_ended'
)


class WebSocketHandler:
    def __init__(self, player):
        self.player = player
        self.socket = None

    async def connect(self, *args):
        try:
            self.socket = await self.player.client.ws_connect(
                WS_URI,
                headers=self.player.headers,
                ssl=False
            )

            print(f'Connected to {WS_URI}')
            self.player.aiohttp.loop.create_task(self.run())
        except Exception as e:
            print(f'{Fore.RED}[WS] Failed to connect to ${WS_URI}: {e}')



    async def run(self):
        try:
            async for msg in self.socket:
                if msg.type == WSMsgType.ERROR:
                    self.handle_error(msg)
                    break
                elif msg.type == WSMsgType.TEXT:
                    self.handle_message(msg.data)
                elif msg.type == WSMsgType.CLOSING:
                    self.handle_closing(msg)
        finally:
            print(f'[WS] Connection to server was closed')

    async def close(self):
        if self.socket is not None:
            await self.socket.close()

    def handle_message(self, msg):
        payload = json.loads(msg)
        event_type = payload['type']
        data = payload['data']

        if event_type in GAME_MESSAGES:
            print(f'{Fore.CYAN}[SERVER] ({event_type}) {data["msg"]}')
        elif event_type == 'player_error':
            print(f'{Fore.RED}[SERVER] ({event_type}) {data}')
        else:
            print(f'{Fore.RED}[SERVER] ({event_type}) {data}')

    def handle_error(self, msg):
        print(f'{Fore.RED}[WS] Error: {msg}')

    def handle_closing(self, msg):
        print(f'[WS] Connection to server is closing: {msg}')
